pkg/gardenlet/controller/federatedseed/shootstate/shootstate_sync_control.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shootstate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsclientset "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	extensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	extensionslisters "github.com/gardener/gardener/pkg/client/extensions/listers/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// SyncController replicates the extensions' resources states in Shoot's ShootState resource
+type SyncController struct {
+	k8sGardenClient      kubernetes.Interface
+	seedClient           kubernetes.Interface
+	extensionsInformers  extensionsinformers.SharedInformerFactory
+	extensionsSeedClient extensionsclientset.Interface
+
+	context             context.Context
```

Comment:
Context should not be an attribute on structs

http://p.agnihotry.com/post/understanding_the_context_package_in_golang/

> Don’t store contexts in a struct, pass them explicitly in functions, preferably, as the first argument.

Reply:
Yes, we considered that comment from the go docs, however there was no way to pass the already existing context which is created when the gardenlet starts to the SyncControllers' `Run()` function.

From what I see in other packages to avoid having context attributes in structs a new BackgroundContext (for example in the flow logic) or context.TODO() is created/used.

We could either switch to that or pass only a "stop" channel in the struct then create a context from the stop channel with ContextFromStopChannel.

Reply:
I am aware of that, but to achieve this the only way that comes to my mind is to extend/refactor the implementation of the workers and how they are handling the Reconcile Func. We probably can achieve it with closures also. What do you prefer. To create new workers that are passing the context to the reconcile functions or to change the current implementation. Or maybe third option, if you have something else in mind.

Reply:
> I am aware of that, but to achieve this the only way that comes to my mind is to extend/refactor the implementation of the workers and how they are handling the Reconcile Func. We probably can achieve it with closures also. What do you prefer.

Both is fine, I would probably try the one with less effort (probably closure?) and, in a separate PR, change all controllers to use the contextified workers consistently.